flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerNG.java

```diff
@@ -939,8 +968,12 @@ StateAndBoolean updateTaskExecutionState(TaskExecutionStateTransition taskExecut
         @Override
         void onTerminalState(JobStatus jobStatus) {
             Preconditions.checkArgument(jobStatus == JobStatus.CANCELED);
-
-            transitionToState(new WaitingForResources(calculateDesiredResources()));
+            componentMainThreadExecutor.schedule(
+                    () -> {
+                        transitionToState(new WaitingForResources(calculateDesiredResources()));
+                    },
+                    backoffTime,
+                    TimeUnit.MILLISECONDS);
```

Comment:
I think we can only do this transition if we are still in the current state. (e.g. it could happen that the user called `cancel` in between).

Reply:
I should have noticed this myself when I decided not to use `runIfState`. The old rule: If your doing something different, think twice ;)